### PR TITLE
Disable calls to abort() in dlmalloc internal checks

### PR DESCRIPTION
**Pull request by kripken**
There are many of them, and they increase code size significantly in smaller programs - 10% of hello world in wasm, 5% in asm.js!

This does mean that we disable those checks, but I think we should. Default builds shouldn't have code to find bugs in dlmalloc or check for memory corruption.

### CODE REVIEW

**Comment by dschuff** [comment]
Does the asmjs backend support `__builtin_trap`? That's probably a safer replacement, as it's guaranteed to cause a trap (as opposed to builtin_unreachable which can be optimized away). I actually don't know for sure if the wasm backend supports it properly, but we should just lower it to a wasm `unreachable` instruction. I'd be interested to see if that was a size savings over calling abort.

**Comment by kripken** [comment]
The asm.js backend lowers it to an ffi which throws, so yeah, that could be improved. The wasm backend seems to support it as it emits different code for it, but I didn't look beyond the diff. Code sizes:
````
asm2wasm
now 13588
trap ~same (still an ffi call)
unreachable 12240

wasm-backend
now 13457
trap 13443
unreachable 12513
````
So going to trap is helpful by just a few bytes in the wasm backend, as it replaces a call to abort() with an unreachable. Whereas marking the code as unreachable lets LLVM get rid of the check as well, for a large code size win.

I agree that trap is the safer route, but given the code size win possible here, I think we should go for unreachable, since while those assert-style checks are nice, they shouldn't be in a default build - people can make a special build for purposes of debugging dlmalloc and memory corruption issues.

**Comment by juj** [comment]
> This does mean that we disable those checks, but I think we should. Default builds shouldn't have code to find bugs in dlmalloc or check for memory corruption.

Agree that release builds would do not need this. However this change means that users' debug builds will not have these checks either. These aborts have triggered often when doing something wrong in memory management, so I think they are useful that we should not nuke them away behind special tricks. I wonder if we could do a `dlmalloc-debug.bc` and `dlmalloc-release.bc` for this?

**Comment by kripken** [comment]
Done. I had to refactor the dlmalloc code in system_libs.py quite a bit to make this practical, but I think it's better now.

File: tools/system_libs.py [hunk 1]
```diff
@@ -261,32 +261,38 @@ def create_compiler_rt(libname):
...
+  def dlmalloc_name():
+    ret = 'dlmalloc'
+    if shared.Settings.USE_PTHREADS:
+      ret += '_threadsafe'
+    if shared.Settings.EMSCRIPTEN_TRACING:
+      ret += '_tracing'
+    if shared.Settings.SPLIT_MEMORY:
+      ret += '_split'
+    if shared.Settings.DEBUG_LEVEL:
+      ret += '_debug'
```
Comment (juj) [inline tools/system_libs.py]: Looks good, btw, is this the first instance where passing `-g1` or bigger affects codegen? I.e. earlier `-gX` have only affected whitespace, function names and minification?

Comment (kripken) [inline tools/system_libs.py]: I believe so, aside from `-g,-g3,-g4` causing debug info to be included (which affects codegen in some ways).

**Pull request merged**